File: comscore/android/src/main/java/com/theoplayercomscore/integration/ComscoreAnalytics.java
package com.theoplayercomscore.integration;

import android.content.Context;
import android.util.Log;

import com.comscore.Analytics;
import com.comscore.BuildConfig;
import com.comscore.Configuration;
import com.comscore.PublisherConfiguration;
import com.theoplayer.android.api.player.Player;

import java.util.HashMap;
import java.util.Map;

@SuppressWarnings("unused")
public class ComscoreAnalytics {
  private static boolean startedTracking;
  private static ComscoreConfiguration comscoreConfiguration;

  public static void start(ComscoreConfiguration configuration, Context context) {
    if (!startedTracking) {
      if (BuildConfig.DEBUG) {
        Log.i("THEOlog", "calling start");
      }
      comscoreConfiguration = configuration;

      PublisherConfiguration.Builder publisherConfigurationBuilder = new PublisherConfiguration.Builder()
        .publisherId(configuration.getPublisherId())
        .secureTransmission(configuration.isSecureTransmission());

      if (configuration.getUserConsent() == "1" || configuration.getUserConsent() == "0") {
        publisherConfigurationBuilder.persistentLabels(new HashMap<String, String>() {
          {
            put("cs_ucfr", configuration.getUserConsent());
          }
        });
      }

      PublisherConfiguration publisherConfiguration = publisherConfigurationBuilder.build();

      Configuration comscoreConfiguration = Analytics.getConfiguration();

      comscoreConfiguration.addClient(publisherConfiguration);
      comscoreConfiguration.setApplicationName(configuration.getApplicationName());
      if (configuration.isChildDirected()) {
        comscoreConfiguration.enableChildDirectedApplicationMode();
      }

      if (configuration.isDebug()) {
        comscoreConfiguration.enableImplementationValidationMode();
      }

      Analytics.start(context);
      startedTracking = true;
    }
  }

  public void setPersistentLabel(String label, String value) {
    if (startedTracking) {
      ComscoreUtils.notifyHiddenEvent(comscoreConfiguration.getPublisherId(), label, value);
    }
  }

  public void setPersistentLabels(Map<String, String> labels) {
    if (startedTracking) {
      ComscoreUtils.notifyHiddenEvents(comscoreConfiguration.getPublisherId(), labels);
    }
  }

  public static ComscoreStreamingAnalytics createComscoreStreamingAnalytics(Player player, String playerVersion, ComscoreMetaData metadata) throws ComscoreAnalyticsException {
    if (startedTracking) {
      if (BuildConfig.DEBUG) {
        Log.i("THEOlog", "will create CSA object");
      }
      return new ComscoreStreamingAnalytics(player, playerVersion, comscoreConfiguration, metadata);
    } else {
      throw new ComscoreAnalyticsException("start should be called first, can't create the streaming analytics");
    }
  }
}

File: comscore/android/src/main/java/com/theoplayercomscore/integration/ComscoreAnalytics.kt
package com.theoplayercomscore.integration

import android.content.Context
import android.util.Log
import com.comscore.Analytics
import com.comscore.BuildConfig
import com.comscore.PublisherConfiguration
import com.theoplayer.android.api.player.Player

private const val TAG = "ComscoreConnector"

class ComscoreAnalytics {
  private var adapter: ComscoreTHEOplayerAdapter? = null
  private var startedTracking = false

  fun initialize(
    context: Context,
    configuration: ComscoreConfiguration,
    player: Player,
    playerVersion: String,
    metadata: ComscoreMetaData
  ) {
    if (startedTracking) {
      return
    }
    startedTracking = true

    if (BuildConfig.DEBUG) {
      Log.i(TAG, "initialize")
    }

    Analytics.getConfiguration().apply {
      addClient(PublisherConfiguration.Builder()
        .publisherId(configuration.publisherId)
        .secureTransmission(configuration.secureTransmission).apply {
          if (configuration.userConsent === "1" || configuration.userConsent === "0") {
            persistentLabels(hashMapOf("cs_ucfr" to configuration.userConsent))
          }
        }
        .build()
      )
      setApplicationName(configuration.applicationName)
      if (configuration.childDirected) {
        enableChildDirectedApplicationMode()
      }
      if (configuration.debug) {
        enableImplementationValidationMode()
      }
    }

    Analytics.start(context)

    adapter = ComscoreTHEOplayerAdapter(player, playerVersion, configuration, metadata)
  }

  fun destroy() {
    adapter?.destroy()
  }

  fun setPersistentLabel(label: String, value: String) {
    adapter?.setPersistentLabel(label, value)
  }

  fun setPersistentLabels(labels: Map<String, String>) {
    adapter?.setPersistentLabels(labels)
  }

  fun update(metadata: ComscoreMetaData) {
    adapter?.setMedatata(metadata)
  }
}
